# Assignment #2: Data Exploration

In this assignment, you will demonstrate your knowledge of the Python and Pandas skills we've learned so far. These include:

- Getting an overview of your data
- Extracting columns
- Removing duplicates from your data
- Creating a subset of your data by matching a string
- Getting a random sample of your data
- Sorting your data 
- Getting fundamental statistical information of a series, such as the mean and median

You will not need to formulate or answer a research question for this assignment. You simply need to demonstrate your ability to perform specific operations on the data set you have selected. In the next assignment, your midterm, you will create a narrative exploration of the data using many of the same methods, but in such a way as to use narrative along with data analysis to answer a research question.

You may use the data set you imported in assignment #1, or you may choose another data set. That means that you can reuse code you wrote in Assignment #1 to import the data.

In the below notebook, I will ask for you to perform a task. Each task will be followed by one or more blank code cells. Use as many cells as you need to complete the task. If you know some other way to complete the task with Python and/or Pandas that is not the way we learned in class, that is also acceptable. Feel free to use Google searches to refresh your memory as to how to perform each task. You will also likely wish to consult the class notebooks, which can be found in [our class repository on GitHub](https://github.com/sha256rma/foundations-of-data-science).

If your data set does not have a needed form of data for a task, write code that would perform the task if a column with that data existed, and provide an explanation as a comment or markdown cell.

You may receive partial credit for some incomplete or incorrect answers. Please feel free to add comments about your thought process, which will make it more likely that you will receive partial credit.

The last cell is a bonus. You will not be penalized for not attempting or completing it.

-----

Import Pandas and read your data in as a data frame. Assign the data frame to a variable, such as `df`.

In [18]:
import pandas as pd

In [122]:
df = pd.read_csv('~/Desktop/ccny python portfolio/netflix_titles.csv')

In [111]:
df

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

Make the data frame visible by placing the data frame variable in a cell by itself.

In [124]:
print (df)

     show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa 

What columns are in your data frame? Use a function or method that shows all the columns.

In [127]:
df.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

Write Python code to output the number of rows in your data frame.

In [129]:
num_rows = df.shape [0]
print(num_rows)

8807


Write code to output a column from your data frame as a series. (Extract a column.)

In [131]:
column_series = df ['director']
print (column_series)

0       Kirsten Johnson
1                   NaN
2       Julien Leclercq
3                   NaN
4                   NaN
             ...       
8802      David Fincher
8803                NaN
8804    Ruben Fleischer
8805       Peter Hewitt
8806        Mozez Singh
Name: director, Length: 8807, dtype: object


In [139]:
column_series = df ['release_year']
print(column_series)

0       2020
1       2021
2       2021
3       2021
4       2021
        ... 
8802    2007
8803    2018
8804    2009
8805    2006
8806    2015
Name: release_year, Length: 8807, dtype: int64


### Read the example below to **count** duplicates based on specific columns (e.g., 'Name' and 'Age')

Check duplicates based on 'Name' and 'Age' columns (modify the subset for your own columns):

`duplicates_subset = df.duplicated(subset=['Name', 'Age'])`

Count the number of duplicates in the subset:

`num_duplicates_subset = duplicates_subset.sum()`

Display no. of duplicate rows based on your column(s):

`print(f"\nNumber of duplicate rows based on 'Name' and 'Age': {num_duplicates_subset}")`


After reading the above, modify the code snippets to check for duplicate columns in your own data frame

In [165]:
duplicates_subset = df.duplicated(subset=['director', 'release_year'])

In [154]:
num_duplicates_subset = duplicates_subset.sum()

In [174]:
print(f"\nNumber of duplicate rows based on 'director' and 'release_year': {num_duplicates_subset}")


Number of duplicate rows based on 'director' and 'release_year': 2855


If any duplicates exist within a column, remove them with the `.drop_duplicates()` function

In [184]:
df_cleaned = df.drop_duplicates(subset=['director','release_year'])

In [186]:
print(df_cleaned)

     show_id     type                             title  \
0         s1    Movie              Dick Johnson Is Dead   
1         s2  TV Show                     Blood & Water   
2         s3  TV Show                         Ganglands   
5         s6  TV Show                     Midnight Mass   
6         s7    Movie  My Little Pony: A New Generation   
...      ...      ...                               ...   
8801   s8802    Movie                           Zinzana   
8802   s8803    Movie                            Zodiac   
8804   s8805    Movie                        Zombieland   
8805   s8806    Movie                              Zoom   
8806   s8807    Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
1                               NaN   
2                   Julien Leclercq   
5                     Mike Flanagan   
6     Robert Cullen, José Luis Ucha   
...                             ...   
8801              

Write Python code to output the number of rows in your data frame.

In [188]:
num_rows = df.shape [0]
print(f"The number of rows in the DataFrame is: {num_rows}")

The number of rows in the DataFrame is: 8807


Compare the new value with the first time you outputted the number of rows above. Is it more, less, or the same?

- When I first imported the dataset, I started of with 8,807 rows and when I inserted "df_cleaned" in the step before the last one it says 5,952, in that case the number of rows decreased. But, when I inserted "df.shape" it tells me 8,807, which remains the same. (I'm confused as to which one to report.)

Create a subset of the data that matches a specific string in a column. That is, extract all rows of the original data frame that contain a specific string in one of the columns. Save the resulting data frame to a variable. (Add more cells if needed)

In [211]:
subset_df = df[
df['director'].str.contains('NaN', case=True, na=True)
]

In [209]:
print(subset_df)

     show_id     type                                title         director  \
0         s1    Movie                 Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show                        Blood & Water              NaN   
3         s4  TV Show                Jailbirds New Orleans              NaN   
4         s5  TV Show                         Kota Factory              NaN   
10       s11  TV Show  Vendetta: Truth, Lies and The Mafia              NaN   
...      ...      ...                                  ...              ...   
8795   s8796  TV Show                      Yu-Gi-Oh! Arc-V              NaN   
8796   s8797  TV Show                           Yunus Emre              NaN   
8797   s8798  TV Show                            Zak Storm              NaN   
8800   s8801  TV Show                   Zindagi Gulzar Hai              NaN   
8803   s8804  TV Show                          Zombie Dumb              NaN   

                                                   

Output the length of the resulting data frame (the subset of your data for which the conditional was true).

In [219]:
num_rows_subset = subset_df.shape [0]
print(" The number of rows in the subset DataFrame is:",
       num_rows_subset)

The number of rows in the subset DataFrame is: 2634


Output the first ten rows of your data set, the last ten rows of your data set, and a random ten rows of your data set (a sample of your data).

In [253]:
netflix_df.head(10)

show_id     type                             title  \
0      s1    Movie              Dick Johnson Is Dead   
1      s2  TV Show                     Blood & Water   
2      s3  TV Show                         Ganglands   
3      s4  TV Show             Jailbirds New Orleans   
4      s5  TV Show                      Kota Factory   
5      s6  TV Show                     Midnight Mass   
6      s7    Movie  My Little Pony: A New Generation   
7      s8    Movie                           Sankofa   
8      s9  TV Show     The Great British Baking Show   
9     s10    Movie                      The Starling   

                        director  \
0                Kirsten Johnson   
1                            NaN   
2                Julien Leclercq   
3                            NaN   
4                            NaN   
5                  Mike Flanagan   
6  Robert Cullen, José Luis Ucha   
7                   Haile Gerima   
8                Andy Devonshire   
9                 Theodore Melfi   

                                                cast  \
0                                                NaN   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
3                                                NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...   
6  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   
9  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   

                                             country          date_added  \
0                                      United States  September 25, 2021   
1                                       South Africa  September 24, 2021   
2                                                NaN  September 24, 2021   
3                                                NaN  September 24, 2021   
4                                              India  September 24, 2021   
5                                                NaN  September 24, 2021   
6                                                NaN  September 24, 2021   
7  United States, Ghana, Burkina Faso, United Kin...  September 24, 2021   
8                                     United Kingdom  September 24, 2021   
9                                      United States  September 24, 2021   

   release_year rating   duration  \
0          2020  PG-13     90 min   
1          2021  TV-MA  2 Seasons   
2          2021  TV-MA   1 Season   
3          2021  TV-MA   1 Season   
4          2021  TV-MA  2 Seasons   
5          2021  TV-MA   1 Season   
6          2021     PG     91 min   
7          1993  TV-MA    125 min   
8          2021  TV-14  9 Seasons   
9          2021  PG-13    104 min   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   
5                 TV Dramas, TV Horror, TV Mysteries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
8                       British TV Shows, Reality TV   
9                                   Comedies, Dramas   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...  
5  The arrival of a charismatic young priest brin...  
6  Equestria's divided. But a bright-eyed hero be...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented 

In [262]:
netflix_df.tail(10)

show_id     type               title                 director  \
8797   s8798  TV Show           Zak Storm                      NaN   
8798   s8799    Movie            Zed Plus  Chandra Prakash Dwivedi   
8799   s8800    Movie               Zenda           Avadhoot Gupte   
8800   s8801  TV Show  Zindagi Gulzar Hai                      NaN   
8801   s8802    Movie             Zinzana          Majid Al Ansari   
8802   s8803    Movie              Zodiac            David Fincher   
8803   s8804  TV Show         Zombie Dumb                      NaN   
8804   s8805    Movie          Zombieland          Ruben Fleischer   
8805   s8806    Movie                Zoom             Peter Hewitt   
8806   s8807    Movie              Zubaan              Mozez Singh   

                                                   cast  \
8797  Michael Johnston, Jessica Gee-George, Christin...   
8798  Adil Hussain, Mona Singh, K.K. Raina, Sanjay M...   
8799  Santosh Juvekar, Siddharth Chandekar, Sachit P...   
8800  Sanam Saeed, Fawad Khan, Ayesha Omer, Mehreen ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8803                                                NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                            country          date_added  \
8797  United States, France, South Korea, Indonesia  September 13, 2018   
8798                                          India   December 31, 2019   
8799                                          India   February 15, 2018   
8800                                       Pakistan   December 15, 2016   
8801                   United Arab Emirates, Jordan       March 9, 2016   
8802                                  United States   November 20, 2019   
8803                                            NaN        July 1, 2019   
8804                                  United States    November 1, 2019   
8805                                  United States    January 11, 2020   
8806                                          India       March 2, 2019   

      release_year rating   duration  \
8797          2016  TV-Y7  3 Seasons   
8798          2014  TV-MA    131 min   
8799          2009  TV-14    120 min   
8800          2012  TV-PG   1 Season   
8801          2015  TV-MA     96 min   
8802          2007      R    158 min   
8803          2018  TV-Y7  2 Seasons   
8804          2009      R     88 min   
8805          2006     PG     88 min   
8806          2015  TV-14    111 min   

                                              listed_in  \
8797                                           Kids' TV   
8798             Comedies, Dramas, International Movies   
8799                       Dramas, International Movies   
8800  International TV Shows, Romantic TV Shows, TV ...   
8801            Dramas, International Movies, Thrillers   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
8797  Teen surfer Zak Storm is mysteriously transpor...  
8798  A philandering small-town mechanic's political...  
8799  A change in the leadership of a political part...  
8800  Strong-willed, middle-class Kashaf and carefre...  
8801  Recovering alcoholic Talal wakes up inside a s...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...

In [264]:
netflix_df.sample(10)

show_id     type                          title  \
7555   s7556    Movie                          Nails   
2088   s2089    Movie                Unknown Origins   
3714   s3715    Movie                       Unbroken   
6693   s6694  TV Show                     Encerrados   
6276   s6277  TV Show                      Behzat Ç.   
1455   s1456  TV Show  Headspace Guide to Meditation   
6238   s6239    Movie    Barbie Star Light Adventure   
3476   s3477    Movie        The Squid and the Whale   
4075   s4076    Movie                   My Ex & Whys   
2685   s2686    Movie             World Famous Lover   

                        director  \
7555               Dennis Bartok   
2088         David Galán Galindo   
3714   Svati Chakravarty Bhatkal   
6693                         NaN   
6276                         NaN   
1455                         NaN   
6238  Andrew Tan, Michael Goguen   
3476               Noah Baumbach   
4075         Cathy Garcia-Molina   
2685              Kranthi Madhav   

                                                   cast        country  \
7555  Leah McNamara, Shauna Macdonald, Steve Wall, R...        Ireland   
2088  Javier Rey, Verónica Echegui, Brays Efe, Anton...          Spain   
3714                                                NaN          India   
6693  Luis Machín, Maria Merlino, Fabián Vena, Ana C...            NaN   
6276  Erdal Beşikçioğlu, Fatih Artman, İnanç Konukçu...         Turkey   
1455                                    Andy Puddicombe  United States   
6238  Erica Lindbeck, Robbie Daymond, Kimberly Woods...  United States   
3476  Jeff Daniels, Laura Linney, Jesse Eisenberg, O...  United States   
4075  Liza Soberano, Enrique Gil, Joey Marquez, Cai ...    Philippines   
2685  Vijay Deverakonda, Raashi Khanna, Aishwarya Ra...          India   

             date_added  release_year rating  duration  \
7555     March 10, 2018          2017  TV-MA    85 min   
2088    August 28, 2020          2020  TV-MA    97 min   
3714      June 26, 2019          2019  TV-14   109 min   
6693      June 18, 2018          2018  TV-MA  1 Season   
6276       May 19, 2017          2010  TV-MA  1 Season   
1455    January 1, 2021          2021   TV-G  1 Season   
6238    October 1, 2017          2016   TV-Y    79 min   
3476    October 1, 2019          2005      R    81 min   
4075  February 27, 2019          2017  TV-14   121 min   
2685     April 15, 2020          2020  TV-MA   147 min   

                                              listed_in  \
7555                                      Horror Movies   
2088                    International Movies, Thrillers   
3714                Documentaries, International Movies   
6693  International TV Shows, Spanish-Language TV Sh...   
6276  Crime TV Shows, International TV Shows, TV Dramas   
1455                    Docuseries, Science & Nature TV   
6238                           Children & Family Movies   
3476               Comedies, Dramas, Independent Movies   
4075             Comedies, Dramas, International Movies   
2685      Dramas, International Movies, Romantic Movies   

                                            description  
7555  Recently paralyzed in a devastating car accide...  
2088  A very chiseled but somber detective must find...  
3714  After the brutal murders of their loved ones, ...  
6693  This fictional series tells the stories of wha...  
6276  An Ankara homicide detective with a dark perso...  
1455  Headspace takes a friendly, animated look at t...  
6238  A hoverboard-riding princess of the cosmos, Ba...  
3476  This insightful drama looks at the crumbling m...  
4075  Popular blogger Cali is hired for a marketing ...  
2685  When his frustrated girlfriend decides to leav...

## Bonus

You may want to refer to the [Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf).

Sort your data frame by a particular column and output the result.

Pick a column in your data set with numeric data (integers or floats). Output the mean (average) and median of that column.